builtin/common/misc_helpers.lua

```diff
@@ -458,15 +458,28 @@ do
 end
 
 --------------------------------------------------------------------------------
-function table.copy(t, seen)
-	local n = {}
-	seen = seen or {}
-	seen[t] = n
-	for k, v in pairs(t) do
-		n[(type(k) == "table" and (seen[k] or table.copy(k, seen))) or k] =
-			(type(v) == "table" and (seen[v] or table.copy(v, seen))) or v
-	end
-	return n
+
+function table.copy(t, preserve_metatables)
+	local seen = {}
+	local function copy(value)
+		if type(value) ~= "table" then
+			return value
+		end
+		local t = value
+		if seen[t] then
+			return seen[t]
+		end
+		local res = {}
+		seen[t] = res
+		for k, v in pairs(t) do
+			res[copy(k)] = copy(v)
+		end
+		if preserve_metatables then
+			setmetatable(res, getmetatable(t))
+		end
+		return res
+	end
+	return copy(t)
```

Comment:
Note: This allows `table.copy` to also copy primitives, but I consider that a feature. Maybe I should document it?

Reply:
Alternatively add an `assert(type(t) == "table")`, but I could imagine this feature being useful for e.g. copying nil (e.g. if the table is optional).

Reply:
Yep. More generally I think of `table.copy` as a function to copy any "pure Lua" value, be it `nil`, a boolean, number, string or table (which would be recursively copied). I think restricting it arbitrarily to tables doesn't make a lot of sense, hence I'm fine with lifting the restriction.